reportssql/ltfu_report_util_functions.sql

```diff
@@ -0,0 +1,128 @@
+-- getDateMostRecentARVAppointmentThatIsBeforeReportEndDate
+
+DROP FUNCTION IF EXISTS getDateMostRecentARVAppointmentThatIsBeforeReportEndDate;
+
+DELIMITER $$
+CREATE FUNCTION getDateMostRecentARVAppointmentThatIsBeforeReportEndDate(
+    p_patientId INT(11),
+    p_endDate DATE) RETURNS DATE
+    DETERMINISTIC
+BEGIN
+    DECLARE result DATE;
+
+    SELECT DATE(pa.start_date_time) INTO result
+    FROM patient_appointment pa
+    JOIN appointment_service aps ON aps.appointment_service_id = pa.appointment_service_id AND aps.voided = 0
+    WHERE pa.voided = 0
+        AND pa.patient_id = p_patientId
+        AND pa.start_date_time < p_endDate
+        AND (
+            aps.name = "APPOINTMENT_SERVICE_ANC_KEY" OR
+            aps.name = "APPOINTMENT_SERVICE_ART_KEY" OR
+            aps.name = "APPOINTMENT_SERVICE_ART_DISPENSARY_KEY" OR
+            aps.name = "APPOINTMENT_SERVICE_OPD_KEY")
+    ORDER BY pa.start_date_time DESC
+    LIMIT 1;
+
+    RETURN (result);
+END$$
+DELIMITER ;
+
+-- getDateMostRecentHIVRelatedEncounterWithinReportingPeriod
+
+DROP FUNCTION IF EXISTS getDateMostRecentHIVRelatedEncounterWithinReportingPeriod;
+
+DELIMITER $$
+CREATE FUNCTION getDateMostRecentHIVRelatedEncounterWithinReportingPeriod(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS DATE
+    DETERMINISTIC
+BEGIN
+    DECLARE result DATE;
+
+    SELECT e.encounter_datetime INTO result
+    FROM encounter e
+    JOIN `location` loc ON loc.location_id = e.location_id
+    WHERE e.voided = 0
+        AND e.patient_id = p_patientId
+        AND e.encounter_datetime BETWEEN p_startDate AND p_endDate
+        AND loc.name IN (
+            "LOCATION_ANC",
+            "LOCATION_ART",
+            "LOCATION_ART_DISPENSATION",
+            "LOCATION_OPD"
+        )
+    ORDER BY e.encounter_datetime DESC
+    LIMIT 1;
+
+    RETURN (result);
+END$$
+DELIMITER ;
+
+-- getDateMostRecentARVPickupWithinReportingPeriod
+
+DROP FUNCTION IF EXISTS getDateMostRecentARVPickupWithinReportingPeriod;
+
+DELIMITER $$
+CREATE FUNCTION getDateMostRecentARVPickupWithinReportingPeriod(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS DATE
+    DETERMINISTIC
+BEGIN
+    DECLARE result DATE;
+
+    SELECT o.scheduled_date INTO result
+    FROM orders o
+    JOIN drug_order do ON do.order_id = o.order_id
+    JOIN concept c ON do.duration_units = c.concept_id AND c.retired = 0
```

Comment:
@tmvumbi2 This join is never used? **JOIN concept c ON do.duration_units = c.concept_id AND c.retired = 0**